sherpa/ui/utils.py

```diff
@@ -2156,8 +2157,13 @@ def set_stat(self, stat):
            Chi-squared with model amplitude variance.
 
         chi2xspecvar
-           Chi-squared with data variance (XSPEC-style, variance = 1.0
-           if data less than or equal to 0.0).
+           Chi-squared with data variance to match XSPEC. Errors from
+           zero-count channels (source or background) are ignored if
```

Comment:
Are you really ignoring the errors in zero-count channels? There are non-zero errors in these channels when I run it with this branch build.

```
In [92]: subtract()

In [93]: calc_stat_info()
Dataset               = 1
Statistic             = chi2xspecvar
Fit statistic value   = 120.547
Data points           = 136
Degrees of freedom    = 134
Probability [Q-value] = 0.791007
Reduced statistic     = 0.899604

In [94]: unsubtract()

In [95]: calc_stat_info()
WARNING: data set 1 has associated backgrounds, but they have not been subtracted, nor have background models been set
Dataset               = 1
Statistic             = chi2xspecvar
Fit statistic value   = 115.964
Data points           = 136
Degrees of freedom    = 134
Probability [Q-value] = 0.867406
Reduced statistic     = 0.865404
```
Is the errors in the plot_data() calculated for the zero counts channel? Maybe that's what confuses me, as I have been looking at get_data_plot().yerr which changes depending on the subtraction. It is 1 for the non-subtracted data when using `chi2xspecvar`

Reply:
It does (at least when the other component is not zero). If it didn't, I wouldn't have asked for review because it would have been wrong.

I don't understand what the output you included here means.

Reply:
You should be able to use `get_staterror` to look at the errors, although you probably want to add `filter=True`

```
sherpa-4.14.0> load_pha("sherpa-test-data/sherpatest/qso.pi")
read ARF file sherpa-test-data/sherpatest/qso.warf
read RMF file sherpa-test-data/sherpatest/qso.wrmf
read ARF (background) file sherpa-test-data/sherpatest/qso_bkg.warf
read RMF (background) file sherpa-test-data/sherpatest/qso_bkg.wrmf
read background file sherpa-test-data/sherpatest/qso_bkg.pi

sherpa-4.14.0> notice(5, 6)

sherpa-4.14.0> set_stat("chi2xspecvar")

sherpa-4.14.0> get_staterror(filter=True)
array([1.41421356, 1.41421356, 2.        , 1.41421356, 2.44948974,
       2.        , 1.73205081, 1.41421356, 2.        , 1.41421356,
       2.44948974, 1.41421356, 1.73205081, 1.        , 1.41421356,
       1.73205081, 1.        , 1.73205081, 2.64575131, 1.41421356,
       1.41421356, 1.41421356, 1.41421356, 2.        , 2.        ,
       1.        , 1.        , 1.41421356, 1.        , 1.41421356,
       2.44948974, 2.23606798, 1.        , 1.41421356, 1.41421356,
       1.        , 1.41421356, 1.73205081, 2.        , 1.73205081,
       2.        , 1.73205081, 2.        , 1.        , 1.        ,
       1.        , 1.        , 1.41421356, 2.23606798, 1.        ,
       1.41421356, 1.41421356, 2.23606798, 2.        , 1.41421356,
       1.41421356, 1.41421356, 1.        , 2.        , 1.41421356,
       1.41421356, 1.41421356, 1.41421356, 2.        , 1.73205081,
       1.73205081, 1.        , 1.41421356, 2.23606798])

sherpa-4.14.0> get_staterror(filter=True, bkg_id=1)
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1.])
```

You can compare to the number of counts in the source and background components with `get_dep`

```
sherpa-4.14.0> get_dep(filter=True)
array([2., 2., 4., 2., 6., 4., 3., 2., 4., 2., 6., 2., 3., 0., 2., 3., 0.,
       3., 7., 2., 2., 2., 2., 4., 4., 1., 1., 2., 1., 2., 6., 5., 0., 2.,
       2., 1., 2., 3., 4., 3., 4., 3., 4., 1., 1., 1., 1., 2., 5., 0., 2.,
       2., 5., 4., 2., 2., 2., 1., 4., 2., 2., 2., 2., 4., 3., 3., 1., 2.,
       5.])

sherpa-4.14.0> get_dep(filter=True, bkg_id=1)
array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0.])
```

You can see here the background is either 0 or 1 counts and chi2xspecvar returns an error of 1 for all of them (which is what it's meant to do).

You can subtract and see the result

```
sherpa-4.14.0> subtract()

sherpa-4.14.0> get_staterror(filter=True)
array([1.43782764, 1.43782764, 2.0167668 , 1.43782764, 2.4631988 ,
       2.0167668 , 1.75138469, 1.43782764, 2.0167668 , 1.43782764,
       2.4631988 , 1.43782764, 1.75138469, 1.03312551, 1.43782764,
       1.75138469, 1.03312551, 1.75138469, 2.65844848, 1.43782764,
       1.43782764, 1.43782764, 1.43782764, 2.0167668 , 2.0167668 ,
       1.03312551, 1.03312551, 1.43782764, 1.03312551, 1.43782764,
       2.4631988 , 2.25107715, 1.03312551, 1.43782764, 1.43782764,
       1.03312551, 1.43782764, 1.75138469, 2.0167668 , 1.75138469,
       2.0167668 , 1.75138469, 2.0167668 , 1.03312551, 1.03312551,
       1.03312551, 1.03312551, 1.43782764, 2.25107715, 1.03312551,
       1.43782764, 1.43782764, 2.25107715, 2.0167668 , 1.43782764,
       1.43782764, 1.43782764, 1.03312551, 2.0167668 , 1.43782764,
       1.43782764, 1.43782764, 1.43782764, 2.0167668 , 1.75138469,
       1.75138469, 1.03312551, 1.43782764, 2.25107715])
```

If I use this PR then I get (the `get_staterror` values should be the same before the `subtract` call but I haven't checked)

```
>>> load_pha("sherpa-test-data/sherpatest/qso.pi")
read ARF file sherpa-test-data/sherpatest/qso.warf
read RMF file sherpa-test-data/sherpatest/qso.wrmf
read ARF (background) file sherpa-test-data/sherpatest/qso_bkg.warf
read RMF (background) file sherpa-test-data/sherpatest/qso_bkg.wrmf
read background file sherpa-test-data/sherpatest/qso_bkg.pi
>>> notice(5, 6)
>>> set_stat("chi2xspecvar")
>>> get_staterror(filter=True)
array([1.41421356, 1.41421356, 2.        , 1.41421356, 2.44948974,
       2.        , 1.73205081, 1.41421356, 2.        , 1.41421356,
       2.44948974, 1.41421356, 1.73205081, 1.        , 1.41421356,
       1.73205081, 1.        , 1.73205081, 2.64575131, 1.41421356,
       1.41421356, 1.41421356, 1.41421356, 2.        , 2.        ,
       1.        , 1.        , 1.41421356, 1.        , 1.41421356,
       2.44948974, 2.23606798, 1.        , 1.41421356, 1.41421356,
       1.        , 1.41421356, 1.73205081, 2.        , 1.73205081,
       2.        , 1.73205081, 2.        , 1.        , 1.        ,
       1.        , 1.        , 1.41421356, 2.23606798, 1.        ,
       1.41421356, 1.41421356, 2.23606798, 2.        , 1.41421356,
       1.41421356, 1.41421356, 1.        , 2.        , 1.41421356,
       1.41421356, 1.41421356, 1.41421356, 2.        , 1.73205081,
       1.73205081, 1.        , 1.41421356, 2.23606798])
>>> get_staterror(filter=True, bkg_id=1)
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1.])
>>> subtract()
>>> get_staterror(filter=True)
array([1.41421356, 1.41421356, 2.        , 1.41421356, 2.44948974,
       2.        , 1.73205081, 1.41421356, 2.        , 1.41421356,
       2.44948974, 1.41421356, 1.73205081, 0.25951556, 1.41421356,
       1.73205081, 0.25951556, 1.73205081, 2.64575131, 1.41421356,
       1.41421356, 1.41421356, 1.41421356, 2.        , 2.        ,
       1.        , 1.        , 1.41421356, 1.        , 1.41421356,
       2.44948974, 2.23606798, 0.25951556, 1.41421356, 1.41421356,
       1.        , 1.41421356, 1.73205081, 2.        , 1.73205081,
       2.        , 1.73205081, 2.0167668 , 1.03312551, 1.        ,
       1.        , 1.        , 1.41421356, 2.23606798, 0.25951556,
       1.41421356, 1.41421356, 2.23606798, 2.        , 1.41421356,
       1.41421356, 1.41421356, 1.        , 2.        , 1.41421356,
       1.41421356, 1.41421356, 1.41421356, 2.        , 1.73205081,
       1.73205081, 1.        , 1.41421356, 2.23606798])
>>> get_staterror(filter=True, bkg_id=1)
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1.])
```

(note that the background errors haven't changed, it's only the source errors that change). You can see they are smaller than the 4.14 versions. You'll notice that the errors are the same as the before-subtraction errors apart for the three bins which have 1 count in the background and the three bins which have 0's in both source and background

In 4.14 we can look at the ratio of y2 / y1 (after background sbutraction to before subtraction) which is always > 1

```
sherpa-4.14.0> unsubtract()

sherpa-4.14.0> y1 = get_staterror(filter=True).copy()

sherpa-4.14.0> subtract()

sherpa-4.14.0> y2 = get_staterror(filter=True).copy()

sherpa-4.14.0> unsubtract()

sherpa-4.14.0> y2 / y1
array([1.01669768, 1.01669768, 1.0083834 , 1.01669768, 1.0055967 ,
       1.0083834 , 1.01116242, 1.01669768, 1.0083834 , 1.01669768,
       1.0055967 , 1.01669768, 1.01116242, 1.03312551, 1.01669768,
       1.01116242, 1.03312551, 1.01116242, 1.00479908, 1.01669768,
       1.01669768, 1.01669768, 1.01669768, 1.0083834 , 1.0083834 ,
       1.03312551, 1.03312551, 1.01669768, 1.03312551, 1.01669768,
       1.0055967 , 1.00671231, 1.03312551, 1.01669768, 1.01669768,
       1.03312551, 1.01669768, 1.01116242, 1.0083834 , 1.01116242,
       1.0083834 , 1.01116242, 1.0083834 , 1.03312551, 1.03312551,
       1.03312551, 1.03312551, 1.01669768, 1.00671231, 1.03312551,
       1.01669768, 1.01669768, 1.00671231, 1.0083834 , 1.01669768,
       1.01669768, 1.01669768, 1.03312551, 1.0083834 , 1.01669768,
       1.01669768, 1.01669768, 1.01669768, 1.0083834 , 1.01116242,
       1.01116242, 1.03312551, 1.01669768, 1.00671231])
```

We now have the interesting case where the ratio is mostly 1, but there are a few slightly larger than 1 and four actually less than 1

```
>>> unsubtract()
>>> y1 = get_staterror(filter=True).copy()
>>> subtract()
>>> y2 = get_staterror(filter=True).copy()
>>> unsubtract()
>>> y2 / y1
array([1.        , 1.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 1.        , 0.25951556, 1.        ,
       1.        , 0.25951556, 1.        , 1.        , 1.        ,
       1.        , 1.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 0.25951556, 1.        , 1.        ,
       1.        , 1.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 1.0083834 , 1.03312551, 1.        ,
       1.        , 1.        , 1.        , 1.        , 0.25951556,
       1.        , 1.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 1.        , 1.        ])
```